os/drivers/input/ist415.c

```diff
@@ -0,0 +1,413 @@
+ 
+/* IST415 Device */
+struct ist415_dev_s
+{
+	/* I2C bus and address for device. */
+	struct i2c_dev_s *i2c;
+	struct i2c_config_s config;
+	uint8_t addr;
+	struct lcd_touch_config *lower;
+
+	/* Configuration for device. */
+	sem_t sem;
+	uint8_t crefs;
+	bool int_pending;
+#ifndef CONFIG_DISABLE_POLL
+	sem_t pollsem;
+	struct pollfd *fds[CONFIG_TOUCH_NPOLLWAITERS];
+#endif
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int ist415_open(FAR struct file *filep);
+static int ist415_close(FAR struct file *filep);
+static ssize_t ist415_read(FAR struct file *filep, FAR char *buffer, size_t buflen);
+static ssize_t ist415_write(FAR struct file *filep, FAR char *buffer, size_t buflen);
+#ifndef CONFIG_DISABLE_POLL
+static void ist415_pollnotify(FAR struct ist415_dev_s *dev, pollevent_t eventset);
+static int ist415_poll(FAR struct file *filep, FAR struct pollfd *fds, bool setup);
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static FAR struct ist415_dev_s g_ist415_priv;
+
+/* File Operations exposed to NuttX Apps */
+static const struct file_operations g_ist415_fileops =
+{
+	ist415_open,   /* open */
+	ist415_close,  /* close */
+	ist415_read,   /* read */
+	ist415_write,           /* write */
+	NULL,           /* seek */
+	NULL,          /*ioctl */
+	ist415_poll,          
+};
+
+/* Last event, last ID and last valid touch coordinates */
+static uint8_t  last_event = 0xff;
+static uint8_t  last_id    = 0xff;
+static uint16_t last_x     = 0xffff;
+static uint16_t last_y     = 0xffff;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static int ist415_get_touch_data(struct ist415_dev_s *dev, FAR void *buf)
+{
+	struct i2c_dev_s *i2c = dev->i2c;
+	struct i2c_config_s config = dev->config;
+	struct ts_event_coordinate *p_evt_coord;
+	struct touch_sample_s *data = buf;
+	u8 event[EVENT_PACKET_SIZE];
+	u8 *left_event;
+	int reg[2] = {HCOM_GET_EVENT_1ST, 0};
+	int ret = i2c_write(i2c, &config, (uint8_t *)reg, 1);
+	if(ret == 1) {
+		i2c_read(i2c, &config, (uint8_t *)event, EVENT_PACKET_SIZE);
+		u8 le  = event[7] & 0x1F;
+		u8 eid = event[0] & 0x3;
+		if (eid == EID_COORD) {
+			p_evt_coord = (struct ts_event_coordinate *) event;
+			data->npoints = 1;
+  			data->point[0].id = p_evt_coord->tid - 1;
+  			data->point[0].x  = (p_evt_coord->x_11_4 << 4) | p_evt_coord->x_3_0;
+  			data->point[0].y  = (p_evt_coord->y_11_4 << 4) | p_evt_coord->y_3_0;
+  			data->point[0].h = 0;
+  			data->point[0].w = 0;
+  			data->point[0].pressure = 0;
+			switch (p_evt_coord->tsta) {
+					data->point[0].flags = TOUCH_DOWN;
+					break;
```

Comment:
1. It looks some case is missing.
2. coding rule. switch and case should have the same indentation.